Area51/Sources/FeedViewController.swift

```diff
@@ -3,70 +3,89 @@ import SafariServices
 import UIKit
 
 final class FeedViewController: UIViewController {
-    @IBOutlet weak var tableView: UITableView!
-    private var dataSource: ListingsDataSource! {
-        didSet {
-            self.dataSource.updated = { [weak self] in
-                self?.tableView.refreshControl?.endRefreshing()
-                self?.tableView.reloadData()
-            }
-        }
-    }
+
+    @IBOutlet private weak var tableView: UITableView!
+    private var tableViewDatasource: TableViewDataSource<Listing>?
+    private var dataSource: ListingsDataSource!
 
     class func instantiateFromStoryboard(withSubreddit subreddit: Subreddit) -> FeedViewController {
         let storyboard = UIStoryboard(name: "Main", bundle: nil)
         let feedViewController: FeedViewController = storyboard.viewControllerFrom(identifier: "FeedViewController")
+
         feedViewController.dataSource = ListingsDataSource(subreddit: subreddit)
         feedViewController.title = subreddit.name
+
         return feedViewController
     }
 
     override func viewDidLoad() {
         super.viewDidLoad()
-        self.setupTableView()
+        setupTableView()
+        refreshData()
     }
+}
+
+extension FeedViewController {
 
     private func setupTableView() {
-        self.tableView.delegate = self
-        self.tableView.dataSource = self
-        self.tableView.refreshControl = UIRefreshControl()
-        self.tableView.refreshControl?.addTarget(self, action: #selector(self.refreshData), for: .valueChanged)
-        self.tableView.register(UINib(nibName: "ListingTableViewCell", bundle: nil),
-                                forCellReuseIdentifier: "ListingTableViewCell")
-        self.tableView.register(UINib(nibName: "ListingThumbnailTableViewCell", bundle: nil),
-                                forCellReuseIdentifier: "ListingThumbnailTableViewCell")
+        tableView.delegate = self
+        setupTableViewRefreshControl()
+        registerTableViewCells()
+    }
+
+    private func setupTableViewRefreshControl() {
+        tableView.refreshControl = UIRefreshControl()
+        tableView.refreshControl?.addTarget(self, action: #selector(refreshData), for: .valueChanged)
+    }
+
+    private func registerTableViewCells () {
+        tableView.register(ListingTableViewCell.nib(),
+                           forCellReuseIdentifier: ListingTableViewCell.reuseIdentifier())
+    }
+
+    private func loadTableViewDataSource(with listings: [Listing]) {
+        tableViewDatasource = TableViewDataSource.make(for: listings)
+        tableView.dataSource = tableViewDatasource
+        tableView.reloadData()
     }
 
     @objc
     private func refreshData() {
-        self.dataSource.refresh()
+        dataSource.refresh { [weak self] (listings) in
+            if let listings = listings {
```

Comment:
Refresh and load more if needed do the same thing. I think we should go back to a single update block with individual methods for refreshing and loading more which will trigger the update block eventually.